src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,461 @@
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHaveExpRuntimeLookup())
+    {
+
+#ifdef DEBUG
+        for (BasicBlock* block : Blocks())
+        {
+            for (Statement* stmt : block->Statements())
+            {
+                for (GenTree* tree : stmt->TreeList())
+                {
+                    assert(!tree->IsCall() || (tree->IsCall() && !tree->AsCall()->IsExpRuntimeLookup()));
+                }
+            }
+        }
+#endif
+        JITDUMP("Current method doesn't have runtime lookups - bail out.")
+        return result;
+    }
+
+    // Find all calls with GTF_CALL_M_EXP_RUNTIME_LOOKUP flag
+    for (BasicBlock* block : Blocks())
+    {
+    TRAVERSE_BLOCK_AGAIN:
+
+        Statement* prevStmt = nullptr;
+        for (Statement* const stmt : block->Statements())
+        {
+            for (GenTree* const tree : stmt->TreeList())
+            {
+                // We only need calls with IsExpRuntimeLookup() flag
+                if (!tree->IsCall() || !tree->AsCall()->IsExpRuntimeLookup())
+                {
+                    continue;
+                }
+                assert(tree->IsHelperCall());
+
+                JITDUMP("Expanding runtime lookup for [%06d] in " FMT_BB ":\n", dspTreeID(tree), block->bbNum)
+                DISPTREE(tree)
+                JITDUMP("\n")
+
+                GenTreeCall* call = tree->AsCall();
+                call->ClearExpRuntimeLookup();
+                assert(call->gtArgs.CountArgs() == 2);
+
+                if (call->IsTailCall())
+                {
+                    assert(!"Unexpected runtime lookup as a tail call");
+                    continue;
+                }
+
+                // call(ctx, signature);
+                GenTree* ctxTree = call->gtArgs.GetArgByIndex(0)->GetNode();
+                GenTree* sigTree = call->gtArgs.GetArgByIndex(1)->GetNode();
+
+                void* signature = nullptr;
+                if (sigTree->IsCnsIntOrI())
+                {
+                    signature = (void*)sigTree->AsIntCon()->IconValue();
+                }
+                else
+                {
+                    // signature is not a constant (CSE'd?) - let's see if we can access it via VN
+                    if (vnStore->IsVNConstant(sigTree->gtVNPair.GetLiberal()))
+                    {
+                        signature = (void*)vnStore->CoercedConstantValue<ssize_t>(sigTree->gtVNPair.GetLiberal());
+                    }
+                    else
+                    {
+                        // Technically, it is possible (e.g. it was CSE'd and then VN was erased), but for Debug mode we
+                        // want to catch such cases as we really don't want to emit just a fallback call - it's too slow
+                        assert(!"can't restore signature argument value");
+                        continue;
+                    }
+                }
+                assert(signature != nullptr);
```

Comment:
Can we introduce a new node (or reuse `GT_RUNTIMELOOKUP`) to avoid this fragile logic?

Reply:
Maybe the problem here is really due to the fact that the pass does not properly handle rationalizing the tree containing the runtime lookup.

I think it would be much less fragile to make the pass work generally by doing this rationalization. It would look similar to `SideEffectExtractor` except with an early abort when it finds the runtime lookup and some additional handling for the operands of the runtime lookup.